cps/elf.c

```diff
@@ -0,0 +1,261 @@
+PERBIT(get_section)(struct elf_file *mod, const char *secname,
+	unsigned long *secsize)
+{
+	void *data = mod->data;
+	unsigned long len = mod->len;
+	int conv = mod->conv;
+	unsigned int i;
+	const char *secstrings;
+
+	ElfPERBIT(Ehdr) *hdr = data;
+	ElfPERBIT(Shdr) *sechdrs;
+	ElfPERBIT(Off) e_shoff;
+	ElfPERBIT(Half) e_shnum, e_shstrndx;
+
+	*secsize = 0;
+
+	/* Check whether there's enough room for ELF header. */
+	if (len < sizeof(*hdr))
+		return NULL;
+
+	/* Read in necessary variables from ELF header and fix endianness. */
+	e_shoff = END(hdr->e_shoff, conv);
+	e_shnum = END(hdr->e_shnum, conv);
+	e_shstrndx = END(hdr->e_shstrndx, conv);
+
+	/* Check whether there's enough room for the section headers. */
+	if (len < e_shoff + e_shnum * sizeof(sechdrs[0]))
+		return NULL;
+
+	sechdrs = (ElfPERBIT(Shdr) *)((uint8_t *)hdr + e_shoff);
+
+	/* Check whether there's enough room for section with section names. */
+	if (len < END(sechdrs[e_shstrndx].sh_offset, conv))
+		return NULL;
+
+	/* Fix endianness in section headers if needed. */
+	for (i = 0; i < e_shnum; i++) {
+		sechdrs[i].sh_name = END(sechdrs[i].sh_name, conv);
+		sechdrs[i].sh_offset = END(sechdrs[i].sh_offset, conv);
+		sechdrs[i].sh_size = END(sechdrs[i].sh_size, conv);
+	}
+
+	/* Find symbol table. */
+	secstrings = (const char *)
+		((uint8_t *)hdr + sechdrs[e_shstrndx].sh_offset);
+	for (i = 1; i < e_shnum; i++) {
+		const char *found_secname = secstrings + sechdrs[i].sh_name;
+		if (strcmp(secname, found_secname) == 0) {
+			ElfPERBIT(Off) secoffset = sechdrs[i].sh_offset;
+
+			/* Not enough room in file for this section. */
+			if (len < secoffset + sechdrs[i].sh_size)
+				return NULL;
+
+			*secsize = sechdrs[i].sh_size;
+			return (uint8_t *)hdr + secoffset;
+		}
+	}
+
+	return NULL;
+}
+
+/* Find the next string in an ELF section. */
+static const char *
+next_string(const char *string, unsigned long *secsize)
+{
+	/* Skip non-zero chars. */
+	while (string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	/* Skip any zero padding. */
+	while (!string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	return string;
+}
+
+static int
+load_string(struct elf_file *mod, const char *attr, char *val, size_t val_len)
+{
+	unsigned long size;
+	const char *strings;
+
+	/* Get strings from modinfo section. */
+	strings = PERBIT(get_section)(mod, ".modinfo", &size);
+	if (strings == NULL) {
+		CPS_LOG(ERR, "Unable to get modinfo section from kernel module\n");
+		return -1;
+	}
+
+	if (strings[size - 1] != 0) {
+		CPS_LOG(ERR, "An unterminated string was found at the end of the modinfo section\n");
+		return -1;
+	}
+
+	/* Skip any zero padding. */
+	while (!strings[0]) {
+		strings++;
+		if (size-- <= 1)
+			return -1;
+	}
+
+	while (strings != NULL) {
+		/* Find lines starting with attribute we're looking for. */
+		if (strncmp(strings, attr, strlen(attr)) == 0) {
+			size_t len;
+			char *eq = strchr(strings, '=');
+			if (eq == NULL) {
```

Comment:
Instead of searching for `=`, we should check that `=` is after `attr` to avoid the chance that we get another parameter whose name starts with `attr`.